events/importer/espoo.py

```diff
@@ -18,84 +20,101 @@
 # Per module logger
 logger = logging.getLogger(__name__)
 
+M = TypeVar("M", bound=Model)
+
+
+PreMapper = Annotated[
+    Callable[[str, Any, dict], dict[str, Any]],
+    """
+    Pre field actions are functions that receive arguments:
+    - field_name: str (field that is currently being mapped)
+    - field_data: Any (field data from the origin object)
+    - data: dict (target object data)
+    and return a new dict of fields to be used for the creation of the target object
+    """,
+]
+
+
+PostMapper = Annotated[
+    Callable[[str, Any, dict, M, dict[str, str]], dict[str, Any]],
+    """
+    Post field actions are functions that receive arguments:
+    - field_name: str (field that is currently being mapped)
+    - field_data: Any (field data from the origin object)
+    - data: dict (target object data)
+    - instance: instance of the model being mapped
+    - instance_map: a map of origin object ids to target (local) object ids
+    and return a new dict of fields to be used for the update of the target object
+    """,
+]
+
+
+class EspooImporterError(Exception):
+    pass
+
 
 def parse_jsonld_id(data: dict) -> str:
     if (ld_reference := data.get("@id")) and (origin_id := ld_reference.split("/")[-2]):
         return origin_id
-    raise RuntimeError("Could not parse jsonld id")
+    raise EspooImporterError("Could not parse jsonld id")
 
 
 def parse_jsonld_ids(data: list) -> list[str]:
     return [parse_jsonld_id(d) for d in data]
 
 
-def _get_data(url: str) -> dict:
+def _get_data(url: str, params: Optional[dict[str, str]] = None) -> dict:
     for _ in range(settings.ESPOO_MAX_RETRIES):
-        response = requests.get(url)
+        response = requests.get(url, timeout=settings.ESPOO_TIMEOUT, params=params)
         if response.status_code == 200:
             return response.json()
+        elif response.status_code < 500:
+            raise EspooImporterError(
+                f"ESPOO API returned status code {response.status_code}"
+            )
```

Comment:
Probably shouldn't retry on all 4xx/5xx codes. Retrying on 503 (service unavailable) should be fine, as well as 429 (too many requests). Of course, this is Linked Events we're talking about, so you're mileage may vary,

Reply:
> you shouldn't need to test the retrying

I most certainly do want to have proof that my code will trigger retries as expected regardless of whether I use my own retry logic or external retry logic. Anyway, I still have the HTTPAdapter/Retries code here, so I'll see if there's some reasonable way to test this.

Reply:
No idea what I've thought here with the retries. Feel free to resolve this.

> I most certainly do want to have proof that my code will trigger retries as expected regardless of whether I use my own retry logic or external retry logic. Anyway, I still have the HTTPAdapter/Retries code here, so I'll see if there's some reasonable way to test this.

I'm guessing you're talking about this thread? https://github.com/City-of-Helsinki/linkedevents/pull/718#discussion_r1339780066

Reply:
Yeah my comments here are not terribly relevant at this point: the HTTPAdapter/Retry is baked in and there's also a test that manages to test the retry functionality so I'm happy. It's using the default retry behaviour, so 429  will retry based on the Retry-After.